File: chat-backend/src/main/java/com/smallchat/backend/user/application/usecase/CreateUserUseCase.java
package com.smallchat.backend.user.application.usecase;

import com.smallchat.backend.global.utils.JwtProvider;
import com.smallchat.backend.global.utils.Tokens;
import com.smallchat.backend.user.application.inputport.CreateUserInputPort;
import com.smallchat.backend.user.application.outputport.UserOutputPort;
import com.smallchat.backend.user.domain.model.User;
import com.smallchat.backend.user.domain.model.vo.Password;
import com.smallchat.backend.user.framework.web.dto.CreateUserDto;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class CreateUserUseCase implements CreateUserInputPort {

    private final JwtProvider jwtProvider;
    private final UserOutputPort userOutputPort;

    @Override
    @Transactional
    public CreateUserDto.Response createUser(CreateUserDto.Request request) {
        String id = request.id();
        Password encrypt = Password.encrypt(request.password());
        String nickname = request.nickname();

        User savedUser = userOutputPort.createUser(User.of(nickname, id, encrypt));

        Tokens tokens = jwtProvider.createTokens(savedUser.getUserId(), savedUser.getNickname());
        userOutputPort.saveRefreshToken(savedUser.getUserId(), tokens.refreshToken());
        return new CreateUserDto.Response(tokens);
    }
}

File: chat-backend/src/main/java/com/smallchat/backend/user/application/usecase/CreateUserUseCase.kt
package com.smallchat.backend.user.application.usecase

import com.smallchat.backend.global.domain.auth.JwtProvider
import com.smallchat.backend.user.domain.interfaces.RefreshTokenRepository
import com.smallchat.backend.user.domain.interfaces.UserRepository
import com.smallchat.backend.user.domain.model.RefreshToken
import com.smallchat.backend.user.domain.model.User
import com.smallchat.backend.user.domain.model.vo.Password
import com.smallchat.backend.user.infrastructure.BcryptPasswordEncoder
import com.smallchat.backend.user.interfaces.web.dto.CreateUserDto
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service

@Service
class CreateUserUseCase(
    private val jwtProvider: JwtProvider,
    private val userRepository: UserRepository,
    private val refreshTokenRepository: RefreshTokenRepository,
    private val passwordEncoder: BcryptPasswordEncoder,
) {
    @Transactional
    fun execute(req: CreateUserDto.Request): CreateUserDto.Response {
        val (id, nickname, password) = req
        val hashPwStr = passwordEncoder.encode(password)
        val hashPw = Password(presentPassword = hashPwStr)
        val savedUser = userRepository.save(User(nickname = nickname, loginId = id, password = hashPw))
        val tokens = jwtProvider.createTokens(savedUser.userIdOrThrow, savedUser.nickname)
        refreshTokenRepository.save(RefreshToken(savedUser.userIdOrThrow, tokens.refreshToken))
        return CreateUserDto.Response(tokens)
    }
}
